render/container.go

```diff
@@ -39,85 +39,93 @@ var ContainerRenderer = MakeFilter(
+	for _, n := range inputNodes {
+		for _, ip := range c.toIPs(n) {
+			if _, exists := ipNodes[ip]; exists {
+				// If an IP is shared between multiple nodes, we can't reliably
+				// attribute an connection based on its IP
+				ipNodes[ip] = "" // blank out the mapping so we don't use it
+			} else {
+				ipNodes[ip] = n.ID
+			}
+		}
 	}
+	var ret = make(report.Nodes)
+	var mapped = map[string]string{} // input node ID -> output node ID
 
-	// Nodes without a hostid may be pseudo nodes
-	if _, ok := m.Latest.Lookup(report.HostNodeID); !ok {
-		if externalNode, ok := NewDerivedExternalNode(m, addr, local); ok {
-			return report.Nodes{externalNode.ID: externalNode}
+	// Now look at all the endpoints and see which map to IP nodes
+	for _, m := range endpoints {
+		scope, addr, port, ok := report.ParseEndpointNodeID(m.ID)
+		if !ok {
+			continue
+		}
+		// Nodes without a hostid may be pseudo nodes - if so, pass through to result
+		if _, ok := m.Latest.Lookup(report.HostNodeID); !ok {
+			if id, ok := externalNodeID(m, addr, local); ok {
+				addToResults(m, id, ret, mapped, func() report.Node {
+					return report.MakeNode(id).WithTopology(Pseudo)
+				})
+				continue
+			}
+		}
+		id, found := ipNodes[report.MakeScopedEndpointNodeID(scope, addr, "")]
+		// We also allow for joining on ip:port pairs.  This is useful for
+		// connections to the host IPs which have been port mapped to a
+		// container can only be unambiguously identified with the port.
+		if !found {
+			id, found = ipNodes[report.MakeScopedEndpointNodeID(scope, addr, port)]
+		}
+		if found && id != "" { // not one we blanked out earlier
+			addToResults(m, id, ret, mapped, func() report.Node {
+				return inputNodes[id]
+			})
 		}
 	}
+	// Copy through any unmatched input nodes
+	for _, n := range inputNodes {
+		if _, found := ret[n.ID]; !found {
+			ret[n.ID] = n
+		}
+	}
+	fixupAdjancencies(inputNodes, ret, mapped)
+	fixupAdjancencies(endpoints, ret, mapped)
+	return ret
+}
 
-	// We also allow for joining on ip:port pairs.  This is useful for
-	// connections to the host IPs which have been port mapped to a
-	// container can only be unambiguously identified with the port.
-	// So we need to emit two nodes, for two different cases.
-	id := report.MakeScopedEndpointNodeID(scope, addr, "")
-	idWithPort := report.MakeScopedEndpointNodeID(scope, addr, port)
-	return report.Nodes{
-		id:         NewDerivedNode(id, m).WithTopology(IP),
-		idWithPort: NewDerivedNode(idWithPort, m).WithTopology(IP),
+// Add Node M to the result set ret under id, creating a new result
+// node if not already there, and updating the old-id to new-id mapping
+func addToResults(m report.Node, id string, ret report.Nodes, mapped map[string]string, create func() report.Node) {
+	result, exists := ret[id]
+	if !exists {
+		result = create()
 	}
+	result.Children = result.Children.Add(m)
+	result.Children = result.Children.Merge(m.Children)
+	ret[result.ID] = result
+	mapped[m.ID] = result.ID
```

Comment:
It is most unfortunate that this function differs from the analogous code in `endpoint2Host.Render`. Do we know these differences are there for good reasons, rather than accidents or even bugs?

Reply:
I've dug into this a bit...

Both here and in `endpoint2Host.Render`, the node we are mapping (`m` in the above) is an endpoint.

I am quite sure endpoints don't have children, so the `result.Children.Merge(m.Children)` here and in `endpoint2Host.Render` are no-ops and superfluous.

I am also quite sure that we do not need to count endpoints. And I certainly cannot think of a reason why we would count them in `endpoint2Host.Render` but not here.

Reply:
I have made this change; dropping the counter doesn't change the final output.

The superfluous `Merge(m.Children)` shouldn't cost much (it detects that case and just returns the input) so I left it; future cases may have children.

Reply:
Now the function is shared, it should be moved near `fixupAdjacencies`.

The function itself is not specific to endpoints, even though the call sites are. I therefore reckon we should add back the counter, or, if that turns out to be really expensive, add a comment. I've spent ages in the past chasing missing counters :(